### PR TITLE
removing Singletons, and just creating new objects when needed

### PR DESCRIPTION
**Pull request by royboy789**
Fix for #5 

@tomjn - let me know what you think.

### CODE REVIEW

File: GutenbergObjectPlugin.php [hunk 1]
```diff
@@ -28,19 +28,7 @@
 use GutenbergArray\API;
 use GutenbergArray\Hooks;
 
-class GutenbergObjectPlugin {
```
Comment (tomjn) [inline GutenbergObjectPlugin.php]: I would keep the gutenberg plugin object, it provides structure and gives all the other objects somewhere to live. Pass the other objects in via the constructor and store them as member variables

File: src/Hooks.php [hunk 2]
```diff
@@ -33,7 +31,7 @@ public function gutes_array_fields() {
 
 	public function get_block_data( $post ) {
 		if ( ! $this->API ) {
-			$this->API = API::init();
+			$this->API = new API();
```
Comment (tomjn) [inline src/Hooks.php]: This leads to multiple API objects, instead eliminate this entire check, and change the constructor

File: src/Hooks.php [hunk 3]
```diff
@@ -6,8 +6,6 @@
 
 class Hooks {
 
-	use Singleton;
-
 	private $API;
 
 	public function __construct() {
```
Comment (tomjn) [inline src/Hooks.php]: Lets use dependency injection by changing the constructor, this object requires an `API` object to do it's work, so lets make it ask for one when it gets created:

```php
public function __construct( API $api ) {
    $this->api = $api;
}
```

Now we don't need to check if api is set, if it isn't set the constructor will throw an error thanks to the type hinting

File: src/Hooks.php [hunk 4]
```diff
@@ -6,8 +6,6 @@
 
 class Hooks {
 
-	use Singleton;
-
 	private $API;
```
Comment (tomjn) [inline src/Hooks.php]: Lets keep variables lowercase, and use uppercase for the Class names instead

**Review by tomjn — CHANGES_REQUESTED** [review]
I'd also recommend moving stuff out of the constructor to a dedicated `run` function. This way creating the object and using the object are separate tasks.

This also means in the root file we get a nice 3 step process:
 - load the class files
 - create and assemble the program from objects like lego
 - tell it we're ready to run and go do things

File: GutenbergObjectPlugin.php (hunk 1 above)
Comment (tomjn) [reply GutenbergObjectPlugin.php]: Keep in mind though that this object should take all the other objects in as dependencies, and it shouldn't do anything itself that isn't a method call of its children.

**Comment by tomjn** [comment]
The final bootstrap might look something like this:

```php

// definitions
namespace GutenbergArray;

define( 'GutesArraySaveUrl', plugin_dir_url( __FILE__ ) );
define( 'GutesArrayPluginFile', __FILE__ );

// load classes
if ( file_exists( __DIR__ . '/vendor/autoload.php' ) ) {
	require_once 'vendor/autoload.php';
}

// construct plugin
$api = new API();
$hooks = new Hooks( $api );
$scripts = new Scripts();
$database = new Database();
$plugin = new GutenbergObjectPlugin( $hooks, $scripts, $database );

// run plugin, register hooks, etc
$plugin->run();
```

Additionally, `GutesArraySaveUrl` and `GutesArrayPluginFile` could be passed in via the constructor too, preventing the need for a `define`, and letting you reuse the class.

If the plugin got bigger, you could swap classes out for interfaces and build different implementations, e.g. `TestDatabase` or `MemcachedDatabase` and as long as all the typehints used the interfaces they could be interchanged. But I don't think that's needed here

**Comment by royboy789** [comment]
@tomjn think I got through all of your notes, and did some more refactoring (namespace was bugging me).

**Comment by tomjn** [comment]
Looks good, my only change would have been to move the top level class to its own file and pass in the  `__FILE__` etc in the constructor so they aren't needed as static variables

**Pull request merged**